docs/available-filters/envelope-encryption.adoc

```diff
@@ -0,0 +1,185 @@
+{kms-api-javadoc}/io/kroxylicious/kms/service/KmsService.html[KMS public API] and make the implementation available on
+the classpath with the service loader mechanism.
+
+== How to use the filter
+
+There are three steps to using the filter.
+
+1. Setting up the KMS.
+2. Configuring the filter within Kroxylicious.
+3. Establishing the encryption key(s) within the KMS that will be used to encrypt the topics.
+
+These steps are described in the next sections.
+
+=== Setting up the KMS
+
+In order to set up the KMS provider ready to use with the filter, follow these KMS provider specific steps.
+
+==== HashiCorp Vault
+
+The filter integrates with the https://developer.hashicorp.com/vault/docs/secrets/transit[HashiCorp Vault *transit
+engine*].   Vault does not enable the transit engine by default.  It must be
+https://developer.hashicorp.com/vault/docs/secrets/transit#setup[enabled] before it can be used with the filter.
+
+The transit engine's path must be `/transit` (the default).
+
+The filter's configuration requires:
+
+* the HashiCorp Vault endpoint
+* a https://developer.hashicorp.com/vault/docs/concepts/tokens[token] that will be used by the filter to read the
+  encryption keys.
+
+=== Filter Configuration
+
+The filter is configured as part of the filter chain in the following way:
+
+[source, yaml]
+----
+filters:
+- type: EnvelopeEncryption                                        # <1>
+  config:
+    kms: KmsServiceName                                           # <2>
+    kmsConfig:                                                    # <3>
+      ..:
+    selector: SelectorName                                        # <4>
+    selectorConfig:                                               # <5>
+      ..:
+----
+<1> The name of the filter. This must be `EnvelopeEncryption`.
+<2> The KMS provider name.
+<3> Object providing configuration understood by KMS provider.
+<4> The key selector name.
+<5> Object providing configuration understood by key selector.
+
+==== KMS configuration
+
+For the KMS configuration:
+
+===== HashiCorp Vault
+
+For HashiCorp Vault, the KMS configuration looks like this:
+
+[source, yaml]
+----
+kms: VaultKmsService                                          # <1>
+kmsConfig:
+  vaultUrl: https://myvault:8200                              # <2>
+  tls:                                                        # <3>
+  vaultToken: mytoken                                         # <4>
+----
+<1> Name of the KMS provider. This must be `VaultKmsService`.
+<2> Vault URL
+<3> (Optional) TLS trust configuration.
+<4> Vault Token
+
+For TLS trust configuration, the filter accepts the same trust parameters as link:../deploying.adoc#_upstream_tls[Upstream TLS]
+except the `PEM` store type is currently https://github.com/kroxylicious/kroxylicious/issues/933[not supported].
+
+==== Key Encryption Key (KEK) selector configuration
```

Comment:
Similar to what Tom was saying, talking about Key Encryption Key without defining it earlier (probably alongside defining DEK and a little more about how envelope encryption works) will be confusing. It's a funny name to see in isolation.

Maybe it needs to start with a `Concepts` section or `What is Envelope Encryption` before describing how to implement it with the filter.